av_metrics/src/capi.rs

```diff
@@ -436,6 +443,327 @@ pub unsafe extern fn avm_calculate_frame_ciede(
+    chroma_pos: ChromaSamplePosition,
+    subsampling: ChromaSampling,
+    pixel_aspect_ratio: Rational,
+    metric: &str,
+) -> (*const Context, f64) {
+    let (_cw, ch) = subsampling.get_chroma_dimensions(width as usize, height as usize);
+
+    let luma_len1 = (frame1_strides[0] as usize) * (height as usize);
+    let luma_slice1 = slice::from_raw_parts(frame1[0], luma_len1);
+    let chroma_u_len1 = (frame1_strides[1] as usize) * (ch as usize);
+    let chroma_u_slice1 = slice::from_raw_parts(frame1[1], chroma_u_len1);
+    let chroma_v_len1 = (frame1_strides[2] as usize) * (ch as usize);
+    let chroma_v_slice1 = slice::from_raw_parts(frame1[2], chroma_v_len1);
+
+    let luma_len2 = (frame2_strides[0] as usize) * (height as usize);
+    let luma_slice2 = slice::from_raw_parts(frame2[0], luma_len2);
+    let chroma_u_len2 = (frame2_strides[1] as usize) * (ch as usize);
+    let chroma_u_slice2 = slice::from_raw_parts(frame2[1], chroma_u_len2);
+    let chroma_v_len2 = (frame2_strides[2] as usize) * (ch as usize);
+    let chroma_v_slice2 = slice::from_raw_parts(frame2[2], chroma_v_len2);
+
+    if bitdepth == 8 {
+        calculate_frame_buf_tmpl::<u8>(
+            [&luma_slice1, &chroma_u_slice1, &chroma_v_slice1],
+            frame1_strides,
+            [&luma_slice2, &chroma_u_slice2, &chroma_v_slice2],
+            frame2_strides,
+            width,
+            height,
+            bitdepth,
+            chroma_pos,
+            subsampling,
+            pixel_aspect_ratio,
+            metric,
+        )
+    } else {
+        calculate_frame_buf_tmpl::<u16>(
+            [&luma_slice1, &chroma_u_slice1, &chroma_v_slice1],
+            frame1_strides,
+            [&luma_slice2, &chroma_u_slice2, &chroma_v_slice2],
+            frame2_strides,
+            width,
+            height,
+            bitdepth,
+            chroma_pos,
+            subsampling,
+            pixel_aspect_ratio,
+            metric,
+        )
+    }
+}
+
+/// Calculate the `ciede` metric between two frame buffers
+///
+/// Returns the correct `ciede` value or `-1` on errors
+#[no_mangle]
+pub unsafe extern fn avm_calculate_frame_buf_ciede(
+    frame1: [*const u8; 3],
+    frame1_strides: [ptrdiff_t; 3],
+    frame2: [*const u8; 3],
+    frame2_strides: [ptrdiff_t; 3],
+    width: u32,
+    height: u32,
+    bitdepth: u8,
+    chroma_pos: ChromaSamplePosition,
+    subsampling: ChromaSampling,
+    pixel_aspect_ratio: Rational,
+) -> f64 {
+    let (_ctx, val) = calculate_frame_buf_internal(
+        frame1,
+        frame1_strides,
+        frame2,
+        frame2_strides,
+        width,
+        height,
+        bitdepth,
+        chroma_pos,
+        subsampling,
+        pixel_aspect_ratio,
+        "ciede",
+    );
+    val
+}
+
+/// Calculate the `ssim` metric between two frame buffers
+///
+/// Returns the correct `ssim` value or `NULL` on errors
+#[no_mangle]
+pub unsafe extern fn avm_calculate_frame_buf_ssim(
+    frame1: [*const u8; 3],
```

Comment:
What cbindgen produces for those?

Reply:
It produces exactly the API I wanted:

```C
/**
 * Calculate the `msssim` metric between two frame buffers
 *
 * Returns the correct `msssim` value or `NULL` on errors
 */
const AVMContext *avm_calculate_frame_buf_msssim(const uint8_t *frame1[3],
                                                 ptrdiff_t frame1_strides[3],
                                                 const uint8_t *frame2[3],
                                                 ptrdiff_t frame2_strides[3],
                                                 uint32_t width,
                                                 uint32_t height,
                                                 uint8_t bitdepth,
                                                 AVMChromaSamplePosition chroma_pos,
                                                 AVMChromaSampling subsampling,
                                                 AVMRational pixel_aspect_ratio);
```

But it doesn't matter, since Rust can't generate a confirming C ABI for it, as discussed on IRC.

Reply:
This is a bug in cbindgen. `ptrdiff_t frame2_strides[3]` is `*mut frame2_strides`.

Reply:
What would be the correct way to write the rust declaration to generate `ptrdiff_t frame2_strides[3]`? And should we file a cbindgen bug?

Reply:
Let me poke upstream to rustc confirm, but from what you are telling me it seems so and a good idea to tell cbindgen upstream.

Reply:
https://github.com/rust-lang/rust/pull/66305 has the background.
I opened https://github.com/eqrion/cbindgen/issues/528 to address it.

Reply:
If it was already part of improper_ctypes, why didn't rustc warn about it during build? That would have been better than finding out via crash.

Reply:
`improper_ctypes` is on by default only in extern blocks.

Reply:
`improper_ctypes` is on by default only in extern blocks.

Reply:
I mean... not's not really a good answer. Why *wouldn't* it be on for all `extern "C"` functions? It makes it almost useless.

Reply:
That's what I had been told today :) I hadn't tested it yet btw.

Reply:
upstream cbindgen would welcome a patch to prevent this and one to add annotations to use the style you'd use here.